### PR TITLE
[✔️] NT-845 Select Reward Button Clicked event

### PR DESCRIPTION
**Pull request by eoji**
# 📲 What
Adding `Select Reward Button Clicked` event.

# 🤔 Why
We have new Back a Project events.

# 🛠 How
- Added `KoalaUtils.pledgeProperties` that takes in a `Reward`
- Added `LakeEvent.SELECT_REWARD_BUTTON_CLICKED` with value `"Select Reward Button Clicked"`
- Added `Koala.trackSelectRewardButtonClicked` that takes in `PledgeData`
- Tracking when a user clicks the `"Select"` button of a `Reward` in an unbacked `Project`
  -  Passing project `RefTag`s into `RewardsFragment`, `BackingFragment` and `PledgeFragment` for tracking.
  - Added `RewardUtils.shippingPreference` for shipping tracking string.
  - `PledgeData` is now an AutoParcel model so its properties can be easily updated as it's passed between screens.
    - Added `PledgeFlowContext` enum with tracking string.
  - Moved `RefTagsAndProject` out of `ProjectViewModel` into its own class called `ProjectData` that takes in a `Project` and its `RefTag`s
  - Renamed `ArgumentsKey.PLEDGE_PROJECT` to `PLEDGE_PROJECT_DATA`. Removed `ArgumentsKey.PLEDGE_PROJECT_DATA`, `PLEDGE_REWARD` and `PLEDGE_SCREEN_LOCATION`.
  - `PledgeFragmentViewModel.OutPuts.startRewardExpandAnimation` now emits a `ScreenLocation`
  - `PledgeFragment` now takes in a `PledgeData` object.
- Renaming configuration input methods to `configureWith`.
- Fixed all the broken tests.
- Added new tests for event and pledge properties.

Added test for PledgeFlowContext in LakeTest.

# 👀 See
No visual changes.

# 📋 QA
So many ways 2 QA:
- `ktk` the staging lake
- check the `Logcat` in Android Studio
- Look at the `dev` project in Amplitude

# Story 📖
[NT-845]

[NT-845]: https://kickstarter.atlassian.net/browse/NT-845

### CODE REVIEW

File: app/src/main/java/com/kickstarter/ui/data/ProjectTracking.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * A light-weight value to hold two ref tags and a project.
+ * Two ref tags are stored: one comes from parceled data in the activity
+ * and the other comes from the ref stored in a cookie associated to the project.
+ */
+
+@AutoParcel
+abstract class ProjectTracking : Parcelable {
+    abstract fun refTagFromIntent(): RefTag?
+    abstract fun refTagFromCookie(): RefTag?
```
[the remark this answers is not in the document]
Comment (eoji) [reply app/src/main/java/com/kickstarter/ui/data/ProjectTracking.kt]: let's add that to the `PledgeData` object instead

File: app/src/main/java/com/kickstarter/viewmodels/BackingFragmentViewModel.kt [hunk 2]
```diff
@@ -31,7 +32,7 @@ interface BackingFragmentViewModel {
         fun pledgeSuccessfullyUpdated()
 
         /** Configure with current project.  */
-        fun project(project: Project)
+        fun project(projectTracking: ProjectTracking)
```
[the remark this answers is not in the document]
Comment (eoji) [reply app/src/main/java/com/kickstarter/viewmodels/BackingFragmentViewModel.kt]: `configureWith`

File: app/src/main/java/com/kickstarter/viewmodels/NativeCheckoutRewardViewHolderViewModel.kt [hunk 3]
```diff
@@ -24,7 +25,7 @@ import kotlin.math.roundToInt
 interface NativeCheckoutRewardViewHolderViewModel {
     interface Inputs {
         /** Call with a reward and project when data is bound to the view.  */
-        fun projectAndReward(project: Project, reward: Reward)
+        fun projectAndReward(projectTracking: ProjectTracking, reward: Reward)
```
[the remark this answers is not in the document]
Comment (eoji) [reply app/src/main/java/com/kickstarter/viewmodels/NativeCheckoutRewardViewHolderViewModel.kt]: `configureWith`

**Comment by eoji** [comment]
I feel good about the changes! @ifbarrera

File: app/src/test/java/com/kickstarter/viewmodels/NativeCheckoutRewardViewHolderViewModelTest.kt [hunk 4]
```diff
@@ -569,54 +569,71 @@ class NativeCheckoutRewardViewHolderViewModelTest : KSRobolectricTestCase() {
...
+        setUpEnvironment(environment())
+
+        this.vm.inputs.configureWith(ProjectDataFactory.project(backedProject), reward)
+        this.showPledgeFragment.assertNoValues()
+
+        // When a reward from a live backed project is clicked, start checkout.
+        this.vm.inputs.rewardClicked(2)
+        this.showPledgeFragment.assertValue(Pair.create(backedProject, reward))
+        this.koalaTest.assertValue("Select Reward Button Clicked")
+        this.lakeTest.assertNoValues()
```
Comment (ifbarrera) [inline app/src/test/java/com/kickstarter/viewmodels/NativeCheckoutRewardViewHolderViewModelTest.kt]: I'm a little confused by this test 🤔 shouldn't the lake client also fire "Select Reward Button Clicked"?

File: app/src/main/java/com/kickstarter/libs/utils/RewardUtils.java [hunk 5]
```diff
@@ -125,6 +125,24 @@ public static boolean isTimeLimited(final @NonNull Reward reward) {
     return context.getString(R.string.discovery_baseball_card_deadline_units_days);
   }
 
+  public static String shippingPreference(final @NonNull Reward reward) {
+    final String shippingType = reward.shippingType();
```
[the remark this answers is not in the document]
Comment (eoji) [reply app/src/main/java/com/kickstarter/libs/utils/RewardUtils.java]: ooh is this the old value? i think i deleted it when we switched to the new format! I can add it back!

File: app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt [hunk 6]
```diff
@@ -576,8 +578,11 @@ interface ProjectViewModel {
                     .compose(bindToLifecycle())
                     .subscribe(this.pledgeActionButtonContainerIsGone)
 
-            currentProjectWhenFeatureEnabled
-                    .filter { it.hasRewards() }
+            val projectTracking = Observable.combineLatest<RefTag, RefTag, Project, ProjectData>(refTag, cookieRefTag, currentProjectWhenFeatureEnabled)
```
[the remark this answers is not in the document]
Comment (eoji) [reply app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt]: yes, thank you!

File: app/src/test/java/com/kickstarter/viewmodels/NativeCheckoutRewardViewHolderViewModelTest.kt (hunk 4 above)
Comment (eoji) [reply app/src/test/java/com/kickstarter/viewmodels/NativeCheckoutRewardViewHolderViewModelTest.kt]: no, i'm only firing for unbacked projects; this test is clicking a reward of a backed project (changing your reward)

File: app/src/main/java/com/kickstarter/libs/Koala.java [hunk 7]
```diff
@@ -751,4 +752,25 @@ public void trackSearchResultsLoaded(final @NonNull DiscoveryParams discoveryPar
     this.client.track(LakeEvent.SEARCH_RESULTS_LOADED, props);
   }
   //endregion
+
+  //region Back a project
+  public void trackSelectRewardButtonClicked(final @NonNull PledgeData pledgeData) {
```
[the remark this answers is not in the document]
Comment (eoji) [reply app/src/main/java/com/kickstarter/libs/Koala.java]: I'm going to unwrap it in the first linne as:
```final ProjectData projectData = pledgeData.projectData();```

File: app/src/main/java/com/kickstarter/viewmodels/BackingFragmentViewModel.kt [hunk 8]
```diff
@@ -111,7 +112,7 @@ interface BackingFragmentViewModel {
     class ViewModel(@NonNull val environment: Environment) : FragmentViewModel<BackingFragment>(environment), Inputs, Outputs {
 
         private val pledgeSuccessfullyCancelled = PublishSubject.create<Void>()
-        private val projectInput = PublishSubject.create<Project>()
+        private val projectInput = PublishSubject.create<ProjectData>()
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/BackingFragmentViewModel.kt]: reame to `projectDataInput`

File: app/src/main/java/com/kickstarter/viewmodels/BackingFragmentViewModel.kt [hunk 9]
```diff
@@ -128,7 +129,7 @@ interface BackingFragmentViewModel {
         private val pledgeDate = BehaviorSubject.create<String>()
         private val pledgeStatusData = BehaviorSubject.create<PledgeStatusData>()
         private val pledgeSummaryIsGone = BehaviorSubject.create<Boolean>()
-        private val projectAndReward = BehaviorSubject.create<Pair<Project, Reward>>()
+        private val projectAndReward = BehaviorSubject.create<Pair<ProjectData, Reward>>()
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/BackingFragmentViewModel.kt]: `projectDataAdReward`

File: app/src/main/java/com/kickstarter/viewmodels/BackingFragmentViewModel.kt [hunk 10]
```diff
@@ -171,8 +173,9 @@ interface BackingFragmentViewModel {
                     .compose(bindToLifecycle())
                     .subscribe(this.backerAvatar)
 
-            backedProject
-                    .map { project -> project.rewards()?.firstOrNull { BackingUtils.isBacked(project, it) }?.let { Pair(project, it) } }
+            this.projectInput
+                    .filter { it.project().isBacking }
+                    .map { projectTracking -> BackingUtils.backedReward(projectTracking.project())?.let { Pair(projectTracking, it) } }
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/BackingFragmentViewModel.kt]: reame to `projectData`

File: app/src/main/java/com/kickstarter/viewmodels/NativeCheckoutRewardViewHolderViewModel.kt [hunk 11]
```diff
@@ -24,7 +27,7 @@ import kotlin.math.roundToInt
 interface NativeCheckoutRewardViewHolderViewModel {
     interface Inputs {
         /** Call with a reward and project when data is bound to the view.  */
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/NativeCheckoutRewardViewHolderViewModel.kt]: use the [Reward] ad [Project]

File: app/src/main/java/com/kickstarter/viewmodels/NativeCheckoutRewardViewHolderViewModel.kt [hunk 12]
```diff
@@ -119,7 +122,7 @@ interface NativeCheckoutRewardViewHolderViewModel {
         private val goRewardlessPreference: BooleanPreferenceType = environment.goRewardlessPreference()
         private val ksCurrency: KSCurrency = environment.ksCurrency()
 
-        private val projectAndReward = PublishSubject.create<Pair<Project, Reward>>()
+        private val projectTrackingAndReward = PublishSubject.create<Pair<ProjectData, Reward>>()
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/NativeCheckoutRewardViewHolderViewModel.kt]: `projectDataAndReward`

File: app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt [hunk 13]
```diff
@@ -609,19 +614,18 @@ interface ProjectViewModel {
...
-                    .map { projectAndReward -> projectAndReward.second?.let { Pair(projectAndReward.first, it) } }
+            val projectDataAndBackedReward = projectData
+                    .map { pD -> BackingUtils.backedReward(pD.project())?.let { Pair(pD, it) } }
 
-            projectAndBackedReward
-                    .compose(takeWhen<Pair<Project, Reward>, Void>(this.updatePaymentClicked))
-                    .map { Pair(PledgeData(reward = it.second, project = it.first), PledgeReason.UPDATE_PAYMENT) }
+            projectDataAndBackedReward
+                    .compose(takeWhen<Pair<ProjectData, Reward>, Void>(this.updatePaymentClicked))
+                    .map { Pair(pledgeData(it.second,  it.first, PledgeFlowContext.MANAGE_REWARD), PledgeReason.UPDATE_PAYMENT) }
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt]: extra whitespace

File: app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt [hunk 14]
```diff
@@ -609,19 +614,18 @@ interface ProjectViewModel {
...
                     .compose(bindToLifecycle())
                     .subscribe(this.showUpdatePledge)
 
-            projectAndBackedReward
-                    .compose(takeWhen<Pair<Project, Reward>, Void>(this.updatePledgeClicked))
-                    .map { Pair(PledgeData(reward = it.second, project = it.first), PledgeReason.UPDATE_PLEDGE) }
+            projectDataAndBackedReward
+                    .compose(takeWhen<Pair<ProjectData, Reward>, Void>(this.updatePledgeClicked))
+                    .map { Pair(pledgeData(it.second,  it.first, PledgeFlowContext.MANAGE_REWARD), PledgeReason.UPDATE_PLEDGE) }
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt]: here too

File: app/src/main/java/com/kickstarter/viewmodels/RewardsFragmentViewModel.kt [hunk 15]
```diff
@@ -41,41 +39,51 @@ class RewardsFragmentViewModel {
 
     class ViewModel(@NonNull val environment: Environment) : FragmentViewModel<RewardsFragment>(environment), Inputs, Outputs {
 
-        private val projectInput = PublishSubject.create<Project>()
+        private val projectTrackingInput = PublishSubject.create<ProjectData>()
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/RewardsFragmentViewModel.kt]: `projectDataInput`

File: app/src/test/java/com/kickstarter/viewmodels/BackingFragmentViewModelTest.kt [hunk 16]
```diff
@@ -513,7 +514,7 @@ class BackingFragmentViewModelTest :  KSRobolectricTestCase() {
 
         setUpEnvironment(environment())
 
-        this.vm.inputs.project(backedProject)
+        this.vm.inputs.configureWith(ProjectDataFactory.project(backedProject))
         this.pledgeSummaryIsGone.assertValue(false)
     }
```
Comment (eoji) [inline app/src/test/java/com/kickstarter/viewmodels/BackingFragmentViewModelTest.kt]: `testProjectDataAndReward`

File: app/src/test/java/com/kickstarter/viewmodels/BackingFragmentViewModelTest.kt [hunk 17]
```diff
@@ -532,8 +533,9 @@ class BackingFragmentViewModelTest :  KSRobolectricTestCase() {
 
         setUpEnvironment(environment())
 
-        this.vm.inputs.project(backedProject)
-        this.projectAndReward.assertValue(Pair(backedProject, reward))
+        val projectTracking = ProjectDataFactory.project(backedProject)
```
Comment (eoji) [inline app/src/test/java/com/kickstarter/viewmodels/BackingFragmentViewModelTest.kt]: `projectData`

File: app/src/test/java/com/kickstarter/viewmodels/RewardsFragmentViewModelTest.kt [hunk 18]
```diff
@@ -47,50 +46,61 @@ class RewardsFragmentViewModelTest: KSRobolectricTestCase() {
...
 
     @Test
-    fun testProject() {
+    fun testProjectData() {
         val project = ProjectFactory.project()
         setUpEnvironment(environment())
 
-        this.vm.inputs.project(project)
-        this.project.assertValue(project)
+        val projectTracking = ProjectDataFactory.project(project)
```
Comment (eoji) [inline app/src/test/java/com/kickstarter/viewmodels/RewardsFragmentViewModelTest.kt]: `projectData`

**Pull request merged**